ui/app/AppLayouts/Chat/views/communities/HoldingsSelectionModel.qml

```diff
@@ -0,0 +1,65 @@
+import SortFilterProxyModel 0.2
+
+import AppLayouts.Chat.helpers 1.0
+import AppLayouts.Chat.panels.communities 1.0
+import AppLayouts.Chat.controls.community 1.0
+
+import StatusQ.Core.Utils 0.1
+
+import utils 1.0
+
+SortFilterProxyModel {
+    property var assetsModel
+    property var collectiblesModel
+
+    proxyRoles: [
+        ExpressionRole {
+            name: "text"
+
+            function getName(type, key) {
+                if (type === HoldingTypes.Type.Ens)
+                    return key
+
+                const model = type === HoldingTypes.Type.Asset
+                            ? assetsModel
+                            : collectiblesModel
+
+                return CommunityPermissionsHelpers.getTokenNameByKey(model, key)
+            }
+
+            function getText(type, key, amount) {
+                const name = getName(type, key)
+
+                return CommunityPermissionsHelpers.setHoldingsTextFormat(
+                            type, name, amount)
+            }
```

Comment:
imho these inner functions shouldn't be inside the `ExpressionRole`, better put them in a private `QtObject`; the _whole_ block gets evaluated by SortFilterProxyModel

Reply:
Are you sure that the stuff outside `expression` property in `ExpressionRole` is treated is some special way? I would guess that's `ExpressionRole` is just an object derived from `QObject` and function defined here is not treated in any special way.

My rationale for putting it here is related to the fact that it's workaround, enforced by limitations/bugs of `expression` property. It's generally better imo when code for workaround is localized possibly close to the source of the problem.

Reply:
https://github.com/oKcerG/SortFilterProxyModel/blob/master/proxyroles/expressionrole.h
https://github.com/oKcerG/SortFilterProxyModel/blob/master/proxyroles/expressionrole.cpp

I haven't analysed it thoroughly but for me only expression is reevaluated by the model so there is no penalty for defining methods inside.

Reply:
Yeah, looks like you're correct; makes sense then